crates/sui-graphql-rpc/src/context_data/db_data_provider.rs

```diff
@@ -509,37 +551,52 @@ impl PgManager {
             .map_err(|e| Error::Internal(e.to_string()))
     }
 
-    /// Takes a query fragment and a lambda that executes the query
+    /// Takes a query_builder_fn that returns Result<QueryFragment> and a lambda to execute the query
     /// Spawns a blocking task that determines the cost of the query fragment
-    /// And if within limits, executes the query
-    async fn run_query_async_with_cost<T, Q, EF, E, F>(
+    /// And if within limits, then executes the query
+    async fn run_query_async_with_cost<T, Q, QResult, EF, E, F>(
         &self,
-        query: Q,
+        mut query_builder_fn: Q,
         execute_fn: EF,
     ) -> Result<T, Error>
     where
-        Q: QueryDsl
+        Q: FnMut() -> Result<QResult, Error> + Send + 'static,
+        QResult: QueryDsl
             + RunQueryDsl<Pg>
             + diesel::query_builder::QueryFragment<diesel::pg::Pg>
+            + diesel::query_builder::Query
+            + diesel::query_builder::QueryId
             + Send
             + 'static,
-        EF: FnOnce(Q) -> F + Send + 'static,
+        EF: FnOnce(QResult) -> F + Send + 'static,
```

Comment:
Can we apply a `Clone` constraint on `Q` instead of requiring that a builder is passed in?

```rust
    async fn run_query_async_with_cost<T, Q, EF, E, F>(
        &self,
        query: Q,
        execute_fn: EF,
    ) -> Result<T, Error>
    where
        Q: QueryDsl
            + RunQueryDsl<Pg>
            + diesel::query_builder::QueryFragment<diesel::pg::Pg>
            + diesel::query_builder::Query
            + diesel::query_builder::QueryId
            + Send
            + Clone
            + 'static,
        EF: FnOnce(QResult) -> F + Send + 'static,
```

Reply:
The issue is that we work with boxed queries, and the limitation w/ Diesel is that once a query is boxed, it cannot be cloned :/ 
We work with boxed queries because of conditional filtering logic

Reply:
An alternative approach would be to use sql string templates

Reply:
https://github.com/diesel-rs/diesel/issues/2277 reason here why we cannot clone